File: third-party/thrift/src/thrift/lib/go/thrift/utils_test.go
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package thrift

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/tls"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"fmt"
	"math/big"
	"net"
	"os"
	"runtime"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

func getNumFileDesciptors(t *testing.T) int {
	// Read the FD directory directly instead of shelling out to lsof: lsof
	// performs blocking kernel/name lookups that can hang for a very long time
	// when the process holds many open network sockets (as TestFDRelease does).
	fdPath := fmt.Sprintf("/proc/%d/fd", os.Getpid())
	if runtime.GOOS == "darwin" {
		fdPath = "/dev/fd"
	}

	fdDir, err := os.Open(fdPath)
	require.NoError(t, err)
	defer fdDir.Close()

	names, err := fdDir.Readdirnames(-1)
	require.NoError(t, err)

	return len(names)
}

func generateSelfSignedCerts() (clientConfig *tls.Config, serverConfig *tls.Config, err error) {
	caPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, nil, err
	}

	caTemplate := x509.Certificate{
		SerialNumber: big.NewInt(1),
		Subject: pkix.Name{
			Organization:  []string{"Test CA"},
			Country:       []string{"US"},
			Province:      []string{""},
			Locality:      []string{"San Francisco"},
			StreetAddress: []string{""},
			PostalCode:    []string{""},
		},
		NotBefore:             time.Now(),
		NotAfter:              time.Now().Add(365 * 24 * time.Hour),
		IsCA:                  true,
		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
		BasicConstraintsValid: true,
	}

	caCertDER, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &caPrivKey.PublicKey, caPrivKey)
	if err != nil {
		return nil, nil, err
	}

	caCert, err := x509.ParseCertificate(caCertDER)
	if err != nil {
		return nil, nil, err
	}

	serverPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, nil, err
	}

	serverTemplate := x509.Certificate{
		SerialNumber: big.NewInt(2),
		Subject: pkix.Name{
			Organization:  []string{"Test Server"},
			Country:       []string{"US"},
			Province:      []string{"CA"},
			Locality:      []string{"Menlo Park"},
			StreetAddress: []string{""},
			PostalCode:    []string{""},
		},
		NotBefore:    time.Now(),
		NotAfter:     time.Now().Add(24 * time.Hour),
		SubjectKeyId: []byte{1, 2, 3, 4, 6},
		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
		KeyUsage:     x509.KeyUsageDigitalSignature,
		IPAddresses:  []net.IP{net.IPv4(127, 0, 0, 1), net.IPv6loopback},
		DNSNames:     []string{"localhost"},
	}

	serverCertDER, err := x509.CreateCertificate(rand.Reader, &serverTemplate, caCert, &serverPrivKey.PublicKey, caPrivKey)
	if err != nil {
		return nil, nil, err
	}

	clientPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, nil, err
	}

	clientTemplate := x509.Certificate{
		SerialNumber: big.NewInt(3),
		Subject: pkix.Name{
			Organization:  []string{"Test Client"},
			Country:       []string{"US"},
			Province:      []string{""},
			Locality:      []string{"Menlo Park"},
			StreetAddress: []string{""},
			PostalCode:    []string{""},
		},
		NotBefore:   time.Now(),
		NotAfter:    time.Now().Add(365 * 24 * time.Hour),
		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
		KeyUsage:    x509.KeyUsageDigitalSignature,
	}

	clientCertDER, err := x509.CreateCertificate(rand.Reader, &clientTemplate, caCert, &clientPrivKey.PublicKey, caPrivKey)
	if err != nil {
		return nil, nil, err
	}

	serverCertPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "CERTIFICATE",
		Bytes: serverCertDER,
	})

	serverKeyPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(serverPrivKey),
	})

	clientCertPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "CERTIFICATE",
		Bytes: clientCertDER,
	})

	clientKeyPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(clientPrivKey),
	})

	caCertPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "CERTIFICATE",
		Bytes: caCertDER,
	})

	serverCert, err := tls.X509KeyPair(serverCertPEM, serverKeyPEM)
	if err != nil {
		return nil, nil, err
	}

	clientCert, err := tls.X509KeyPair(clientCertPEM, clientKeyPEM)
	if err != nil {
		return nil, nil, err
	}

	caCertPool := x509.NewCertPool()
	caCertPool.AppendCertsFromPEM(caCertPEM)

	serverConfig = &tls.Config{
		Certificates: []tls.Certificate{serverCert},
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    caCertPool,
	}

	clientConfig = &tls.Config{
		Certificates: []tls.Certificate{clientCert},
		RootCAs:      caCertPool,
		ServerName:   "localhost",
	}

	return clientConfig, serverConfig, nil
}

func getClientTransportOption(serverTransport TransportID) ClientOption {
	switch serverTransport {
	case TransportIDHeader:
		return withHeader()
	case TransportIDUpgradeToRocket:
		return WithUpgradeToRocket()
	case TransportIDRocket:
		return WithRocket()
	default:
		panic("unsupported transport!")
	}
}

File: third-party/thrift/src/thrift/conformance/UtilsTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <thrift/conformance/Utils.h>

#include <gmock/gmock.h>
#include <gtest/gtest.h>

namespace apache::thrift::conformance {
namespace {
using ::testing::Pair;
using ::testing::UnorderedElementsAre;

TEST(UtilsTest, ParseNameAndCmd) {
  EXPECT_THAT(parseNameAndCmd("foo/bar"), Pair("foo", "foo/bar"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar"), Pair("foo", "/foo/bar"));
  EXPECT_THAT(parseNameAndCmd("/baz/foo/bar"), Pair("foo", "/baz/foo/bar"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar/baz"), Pair("bar", "/foo/bar/baz"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar#"), Pair("foo", "/foo/bar"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar##"), Pair("foo", "/foo/bar#"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar#baz"), Pair("baz", "/foo/bar"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar##baz"), Pair("baz", "/foo/bar#"));
  EXPECT_THAT(parseNameAndCmd("/foo/bar#baz#"), Pair("foo", "/foo/bar#baz"));
  EXPECT_THAT(parseNameAndCmd("/foo#bar/baz"), Pair("foo#bar", "/foo#bar/baz"));
}

TEST(UtilsTest, ParseCmds) {
  auto actual = parseCmds("/foo/bar, /foo/baz#biz ,foo#bar/baz");
  EXPECT_THAT(
      actual,
      UnorderedElementsAre(
          Pair("foo", "/foo/bar"),
          Pair("biz", "/foo/baz"),
          Pair("foo#bar", "foo#bar/baz")));
}

TEST(UtilsTest, ParseCmds_Dupe) {
  EXPECT_THROW(parseCmds("/foo/bar,/baz/foo/bar"), std::invalid_argument);
}

TEST(UtilsTest, ParseNonconforming) {
  EXPECT_THAT(
      parseNonconforming(
          " # Comment\n"
          "test1 # Other comment\n"
          " test2"),
      UnorderedElementsAre("test1", "test2"));
}

TEST(UtilsTest, ParseTestSuite) {}

} // namespace
} // namespace apache::thrift::conformance
